examples/vision/adamatch.py

```diff
@@ -0,0 +1,453 @@
+
+
+def strong_augment(image, source=True):
+    if image.dtype != tf.float32:
+        image = tf.cast(image, tf.float32)
+
+    if source:
+        image = tf.image.resize_with_pad(image, RESIZE_TO, RESIZE_TO)
+        image = tf.tile(image, [1, 1, 3])
+    image = augmenter.distort(image)
+    return image
+
+
+"""
+## Data loader utilities
+"""
+
+
+def create_individual_ds(ds, aug_func, source=True):
+    if source:
+        batch_size = SOURCE_BATCH_SIZE
+    else:
+        # During training 3x more target unlabeled samples are shown
+        # to the model in AdaMatch (Section 3.2 of the paper).
+        batch_size = TARGET_BATCH_SIZE
+    ds = ds.shuffle(batch_size * 10, seed=42)
+
+    if source:
+        ds = ds.map(lambda x, y: (aug_func(x), y), num_parallel_calls=AUTO)
+    else:
+        ds = ds.map(lambda x, y: (aug_func(x, False), y), num_parallel_calls=AUTO)
+
+    ds = ds.batch(batch_size).prefetch(AUTO)
+    return ds
+
+
+"""
+`_w` and `_s` suffixes denote weak and strong respectively.
+"""
+
+source_ds = tf.data.Dataset.from_tensor_slices((mnist_x_train, mnist_y_train))
+source_ds_w = create_individual_ds(source_ds, weak_augment)
+source_ds_s = create_individual_ds(source_ds, strong_augment)
+final_source_ds = tf.data.Dataset.zip((source_ds_w, source_ds_s))
+
+target_ds_w = create_individual_ds(svhn_train, weak_augment, source=False)
+target_ds_s = create_individual_ds(svhn_train, strong_augment, source=False)
+final_target_ds = tf.data.Dataset.zip((target_ds_w, target_ds_s))
+
+"""
+Here's how the image batches look like:
+
+![](https://i.ibb.co/FDVkbWL/Data-Viz.png)
+"""
+
+"""
+## Utilities for loss calculation utilities and weight scheduler
+"""
+
+
+def compute_loss_source(source_labels, logits_source_w, logits_source_s):
+    loss_func = keras.losses.CategoricalCrossentropy(from_logits=True)
+    # First compute the losses between original source labels and
+    # predictions made on the weakly and strongly augmented versions
+    # of the same images.
+    w_loss = loss_func(source_labels, logits_source_w)
+    s_loss = loss_func(source_labels, logits_source_s)
+    return w_loss + s_loss
+
+
+def compute_loss_target(target_pseudo_labels_w, logits_target_s, mask):
+    loss_func = keras.losses.categorical_crossentropy
+    target_pseudo_labels_w = tf.stop_gradient(target_pseudo_labels_w)
+    # For calculating loss for the target samples, we treat the pseudo labels
+    # as the ground-truth. These are not considered during backpropagation
+    # which is a standard SSL practice.
+    target_loss = loss_func(target_pseudo_labels_w, logits_target_s, from_logits=True)
+
+    # More on `mask` later.
+    mask = tf.cast(mask, target_loss.dtype)
+    target_loss *= mask
+    return tf.reduce_mean(target_loss, 0)
+
+
+"""
+Rather than using a fixed scalar quantity, a varying scalar is used in AdaMatch. It
+denotes the weight of the loss contibuted by the target samples.
+"""
+
+
+def mu_schedule():
+    global T
```

Comment:
The use of a global here is an anti-pattern. You should make this function into a model method, and make T (rename it to its full name) into a model property.

Reply:
Sure. The value of `T` represents the current training step. I was wondering how to best achieve that and `global` is what I could think about. Could you expand on this a bit more?

Reply:
I think the current implementation isn't doing what you think it's doing. `T` is a Python variable, so when `train_step` gets compiled it will capture the value of the variable at the time of compilation. It will never get incremented. Here's what you should do:

```python
class AdaMatch(keras.Model):
  def __init__(self, total_steps, ...):
    super(AdaMatch, self).__init__()
    self.total_steps = total_steps
    self.current_step = tf.Variable(0, dtype='int64')
    ...

  def compute_mu(self):
    pi = tf.constant(np.pi, dtype='float32')
    step = tf.cast(self.current_step, dtype='float32')
    return 0.5 - tf.cos(tf.math.minimum(pi, (2 * pi * step) / self.total_steps)) / 2)

  def train_step(self, data):
    ...
    t = self.compute_mu()
    self.current_step.assign_add(1)
    ...
```

Does this make sense?

Reply:
Totally. Thank you.